python/tank/platform/software_launcher.py

```diff
@@ -0,0 +1,738 @@
+            # PROJECT_PATH/tank/config/foo/bar/baz.png
+            config_folder = self.__tk.pipeline_configuration.get_config_location()
+            adjusted_value = value.replace("/", os.path.sep)
+            processed_val = os.path.join(config_folder, adjusted_value)
+
+        elif type(value) == str and value.startswith("hook:"):
+            # handle the special form where the value is computed in a hook.
+            #
+            # if the template parameter is on the form
+            # a) hook:foo_bar
+            # b) hook:foo_bar:testing:testing
+            #
+            # The following hook will be called
+            # a) foo_bar with parameters []
+            # b) foo_bar with parameters [testing, testing]
+            #
+            chunks = value.split(":")
+            hook_name = chunks[1]
+            params = chunks[2:]
+            processed_val = self.__tk.execute_core_hook(
+                hook_name, setting=key, bundle_obj=self, extra_params=params
+            )
+        else:
+            # pass-through
+            processed_val = value
+
+        return processed_val
+
+
+class SoftwareVersion(object):
+    """
+    Container class that stores properties of a DCC that
+    are useful for Toolkit Engine Startup functionality.
+    """
+    def __init__(self, name, version, display_name, path, icon=None):
+        """
+        Constructor.
+
+        :param name: Internal name for the SoftwareVersion
+                     (e.g. Maya)
+        :param version: Explicit (string) version of the DCC represented
+                        (e.g. 2017)
+        :param display_name: Name to use for any graphical displays
+        :param path: Full path to the DCC executable. This path may contain
+                     {version} and/or {v1}..{vN} patterns or ENV variables.
+        :param icon: (Optional) Full path to the icon to use for graphical
+                     displays of this SoftwareVersion. This path may contain
+                     {version} and/or {v1}..{vN} patterns or ENV variables.
+        """
+        self._name = name
+        self._version = version
+        self._display_name = display_name
+        self._path = path
+        self._icon_path = icon
+
+    @property
+    def name(self):
+        """
+        The internal name for this SoftwareVersion
+
+        :returns: String name
+        """
+        return self._name
+
+    @property
+    def version(self):
+        """
+        An explicit version of the DCC represented by this SoftwareVersion.
+
+        :returns: String version
+        """
+        return self._version
+
+    @property
+    def display_name(self):
+        """
+        Name to use for this SoftwareVersion in graphical displays.
+        Replaces any {version} and/or {v1}...{vN} patterns in the
+        display_name string originally specified.
+
+        :returns: String display name
+        """
+        return _apply_version_to_setting(self._display_name, self._version)
```

Comment:
I'm torn about adding the "apply_version_to_setting" functionality to this class at this time.  It is not used at the moment because the launchapp takes care of all the substitution on it's end, but I could see how the functionality might be useful for someone working outside of the launchapp.  Happy to discuss this further.